### PR TITLE
reference-from-sender in inbox-document

### PR DESCRIPTION
**Pull request by runeflobakk**
Lots of file changes for very little behavioral changes. As usual.

Only bizzniz value-add is in f05e28a, where a new optional reference, set by the sender, is made available to the receiver through the inbox API. The library can be released as the added element of the API is optional, and will just yield `Optional.empty()` until Digipost actually supplies the `reference-from-sender` on `inbox-document`s.

Ok, maaaybe some nice-eties in 9fbb52b as well, which should improve the error resolving a little bit. I have no idea why `IllegalStateException | DataBindingException` warrants the construction of an `ErrorMessage` and discarding the stacktrace, but this behavior is kept as-is. I am catching any other exception (not just `IOException`) however, to be able to add some context of the request/response.

Other notable technical improvements:
- The library now don't depend on commons-logging, which it should not have done in the first place.
- logback is removed from test dependencies, use just slf4j-simple instead as logger impl. for tests
- Maven plugins and dependencies upgraded

By the way, the instantiation of the DigipostClient is a f**king disaster. So many randomly overloaded constructors with more or less arbitrary combinations of parameters. But I digress...

### CODE REVIEW

File: src/main/java/no/digipost/api/client/representations/Document.java [hunk 1]
```diff
@@ -125,19 +124,14 @@ public Document(String uuid, String subject, FileType fileType, String openingRe
...
-                cleanedStrings.add(st.trim());
-            }
-        }
-
-        return cleanedStrings.size() != 0 ? StringUtils.join(cleanedStrings, ",") : null;
-
+        return Stream.of(technicalTypes)
+            .filter(s -> Objects.nonNull(s) && !s.isEmpty())
+            .map(String::trim)
+            .collect(collectingAndThen(joining(","), joined -> joined.isEmpty() ? null : joined));
```
Comment (froden) [inline src/main/java/no/digipost/api/client/representations/Document.java]: 🌈

File: src/main/java/no/digipost/api/client/Communicator.java [hunk 2]
```diff
@@ -72,16 +76,35 @@ public static void checkResponse(final CloseableHttpResponse response, EventLogg
...
-            ErrorMessage errorMessage = unmarshal(errorMessageContext, response.getEntity().getContent(), ErrorMessage.class);
+            responseContent = EntityUtils.toByteArray(responseEntity);
+        } catch (IOException e) {
+            throw new DigipostClientException(ErrorCode.GENERAL_ERROR,
+                    "Server returnerte " + statusLine + ", men klienten feilet ved lesing av responsen: " + e.getMessage(), e);
+        }
+
+        try {
+            ErrorMessage errorMessage = unmarshal(errorMessageContext, new ByteArrayInputStream(responseContent), ErrorMessage.class);
             response.close();
```
Comment (froden) [inline src/main/java/no/digipost/api/client/Communicator.java]: Is the response closed somewhere else in case of exception?

**Comment by froden** [comment]
> By the way, the instantiation of the DigipostClient is a f**king disaster. So many randomly overloaded constructors with more or less arbitrary combinations of parameters. But I digress...

Haha, angry @runeflobakk is :rage1: 😄

File: src/main/java/no/digipost/api/client/Communicator.java (hunk 2 above)
Comment (runeflobakk) [inline src/main/java/no/digipost/api/client/Communicator.java]: Oh, I my code never throw exceptions!

Grmf... I'll investigate 😉

Comment (runeflobakk) [reply src/main/java/no/digipost/api/client/Communicator.java]: An `AutoCloseable` _given_ as a method argument should _never_ ble closed by the receiving method. It should be handled where (or very near) the `AutoCloseable` is acquired, as this is where it is known when the resources may be disposed.

Maybe the new ability in Java 9 to just say `try (resourceGotFromSomewhereElse) { .. }` will be an enabler for this antipattern 🙄  

I started to refactor for guarantied response closing, and eventually discovered that I also closed the response when there was no error, and the response is going to be further processed outside this method. And would it be ok to close the response if this method discovers that there is an error? Probably, but using this method requires proper resource handling outside of this method anyway.

File: src/main/java/no/digipost/api/client/util/HttpClientUtils.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+            entity = response.getEntity();
+            return entity.getContent();
+        } catch (IOException | RuntimeException e) {
+            try (CloseableHttpResponse autoClosed = response) {
+                EntityUtils.consume(entity);
+            } catch (IOException | RuntimeException entityConsumptionException) {
+                e.addSuppressed(entityConsumptionException);
+            }
+            throw e instanceof DigipostClientException ? (DigipostClientException) e : new DigipostClientException(GENERAL_ERROR, e.getMessage(), e);
+        }
```
Comment (froden) [inline src/main/java/no/digipost/api/client/util/HttpClientUtils.java]: Sweeeet ❤️

Comment (froden) [reply src/main/java/no/digipost/api/client/util/HttpClientUtils.java]: Referring to the whole method 💯

**Pull request merged**